### PR TITLE
[Research Projects] ORPO diffusion for alignment

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

ORPO `trl`: https://github.com/huggingface/trl/pull/1435
ORPO: [ORPO: Monolithic Preference Optimization without Reference Model](https://arxiv.org/abs/2403.07691)

The PR assumes that MSE in the diffusion formulation approximates the log-probs as required by ORPO. If that breaks but we still get reasonable results, then "more alchemy than science".

### CODE REVIEW

**Comment by sayakpaul** [comment]
@kashif here's a complete short run: https://wandb.ai/sayakpaul/diffusion-orpo-lora-sdxl/runs/s5lufnql. It was run with your small "kashif/pickascore" dataset. 

The results look promising. On the WandB run page, you would also notice that the final inference runs include "test" and "test_without_lora" making the effectivity of this experiment somewhat evident. 

Let me know what you think.

**Comment by sayakpaul** [comment]
@kashif going to merge this but we can continue the experiments we were discussing internally. Hope that's okay.

File: examples/research_projects/diffusion_orpo/train_diffusion_orpo_sdxl_lora.py [hunk 1]
```diff
@@ -0,0 +1,1088 @@
...
+                else:
+                    raise ValueError(f"Unknown prediction type {noise_scheduler.config.prediction_type}")
+
+                # ODDS ratio loss.
+                # In the diffusion formulation, we're assuming that the MSE loss
+                # approximates the logp.
+                model_losses = F.mse_loss(model_pred.float(), target.float(), reduction="none")
+                model_losses = model_losses.mean(dim=list(range(1, len(model_losses.shape))))
+                model_losses_w, model_losses_l = model_losses.chunk(2)
+                log_odds = model_losses_w - model_losses_l
```
Comment (feffy380) [inline examples/research_projects/diffusion_orpo/train_diffusion_orpo_sdxl_lora.py:989]: Is this correct? In language modeling, `log_odds` is positive when the odds of the preferred response are higher than the rejected response. For MSE, this formulation is only positive when the diffusion loss of the preferred image is *worse* than the rejected image.
Shouldn't `log_odds` or `ratio_losses` be negated?
